Signup test: Swap the domain and plan step

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

Check pbKBsr-2c-p2. The screens for the new flow can be seen in [Figma](https://www.figma.com/file/goct7nlLeTqmKMx2cxwQpK/Plans-A%2FB-Test?node-id=115%3A0)

* Swap the domain and plan step in signup
* Launch as A/B test 
* For users who select a free plan followed by a custom domain, we will show an upsell offer right after the domain step:

<img width="858" alt="Screenshot 2020-04-23 at 9 10 28 AM" src="https://user-images.githubusercontent.com/1269602/80056643-4c243b00-8542-11ea-9823-d8fa7063a7a0.png">

Users who click Yes on the offer will see a reduced plans step with only Personal and Premium:

<img width="1005" alt="Screenshot 2020-04-23 at 9 14 13 AM" src="https://user-images.githubusercontent.com/1269602/80056860-d53b7200-8542-11ea-89fe-e48d02272c71.png">


**Sequence of steps for users in the swap flow**

Plans step -> Domains step -> maybe upsell offer(shown only if free plan + custom domain selected) -> maybe a reduced plans step(shown only to users who saw the upsell offer) -> maybe checkout if there is an item in cart -> Customer Home / checklist page.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
For all the scenarios below, please asssign yourself to `variantShowSwapped` of the `domainStepPlanStepSwap` test unless otherwise mentioned.

**Scenario 1**

* Go through the signup flow at /start
* The plans step should appear right after user account creation. Select a free plan.
* You should see the domain step next. Select a free *.wordpress.com domain. 
* Verify that UI matches [Figma](https://www.figma.com/file/goct7nlLeTqmKMx2cxwQpK/Plans-A%2FB-Test?node-id=115%3A0).
* Verify you are taken to Customer Home page.

**Scenario 2**

* Do not yet manually assign yourself to any A/B test.
* Go to /start - you should see the registration page.
* In your browser console, type `localstorage.ABTests` and verify that you are not assigned to swap steps A/B test.
* Try to submit some invalid email combinations and verify that you are still not assigned to the test.
* Complete the user registration successfully. At this point, you should be assigned to the `domainStepPlanStepSwap` test.

**Scenario 3**

* Go through the /start and verify the following combinations:
    - A free plan followed by free domain takes you to Customer Home with your selected free domain
    - A paid plan followed by free domain takes you to checkout page. Completing payment should take you to Customer Home. 
    - A paid plan followed by a custom domain takes you to checkout page. Completing payment should take you to Customer Home. 
    - A free plan followed by a custom domain should take you to the upsell offer page(screenshot in PR description).
        - Click "Yes" on the offer will take you to a reduced plans step with only Personal and Premium plans(check screenshot in description). Selecting a plan will take to checkout.
        - Click "No" on the offer will show the processing screen and then Customer Home.

**Scenario 4**

* In the plans step, select a paid plan and reach the domain step.
* Now hit the Back button in the screen to go back to Plans step.
* Now, select "Start with free" and reach the domain step. 
* Select a custom domain and verify that you see the upsell offer next.

**Scenario 5**

* On an existing account, click Add new site and verify that you get assigned to the A/B test only at the point when the signup flow begins (by typing `localstorage.ABTests` in your browser console and verifying that `domainStepPlanStepSwap` is in the results). You should see the swap flow if in the variation. 

**Scenario 6**

* Verify that the upsell offer copy shows the correct price for domains and plans(verify in a few currencies). The copy talks about pricing in the following places:
"That’s why we’ll cover your domain registration for the first year (saving you ₹1,050), with any paid plan."
"Upgrade to any paid plan, and secure your domain {DOMAIN_NAME}, for just as low as ₹200/mo."

**Scenario 7**

* Selecting a free plan should show the custom domain explainer text in the domain step:

<img width="1031" alt="Screenshot 2020-04-23 at 9 29 26 AM" src="https://user-images.githubusercontent.com/1269602/80057721-f1d8a980-8544-11ea-9591-7073c9adaf64.png">

* Selecting a paid plan should not display any explainer text.

**Scenario 8**
* Verify that when you are in the control group of the swap test, the flow is undisturbed. The domains step, plans step etc. work as before, you are able to complete paid plan purchase or start a free site.

a8ci18n commented on client/my-sites/checkout/upsell-nudge/plan-with-domain/index.jsx:
```diff
@@ -0,0 +1,116 @@
+
+						<p>
+							{ translate(
+								'Give our plans a risk-free test drive with our {{u}}30-day Money Back Guarantee{{/u}}.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Give the Premium plan a risk-free test drive with our {{u}}30-day Money Back Guarantee{{/u}}.' )` *ES Score: 9*

niranjan-uma-shankar commented:
> Scenario 1: I'm seeing the green "Free" in the domains, but not the "Free with paid plan

I fixed this. So you will now see the green "Free" if you select a paid plan in plans step, and the green "Free with paid plan" if you select "Start with free" in the plans step.

> A free plan followed by a custom domain: It's taking a lot of time after selecting a custom domain to move to the next screen, usually up to 5 seconds or more. Also weirdly, although probably not related, in calypso.live I'm only getting on sale .live suggestions, but no other TLD. In localhost I'm getting more domains suggested.

Is the delay reduced when in control group? I notice a delay too, about 2-3 seconds but it's the same in control and variant. Also, looks to be the same in prod too. I'll continue to keep a watch, probably someone can confirm if they see it too. 
I'm guessing that the sale suggestions is probably an environment thing.

a8ci18n commented on client/components/domains/domain-product-price/index.jsx:
```diff
@@ -46,14 +51,30 @@ class DomainProductPrice extends React.Component {
+				if ( isEligibleVariantForDomainTest ) {
+					if ( showFreeDomainExplainerForFreePlan ) {
+						message = translate(
+							'Registration fee: {{del}}%(cost)s{{/del}} {{span}}Free with paid plan{{/span}}',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( '{{del}}%(cost)s{{/del}} {{span}}Free with a paid plan{{/span}}' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/swap-domain-plan-abtest)

p-jackson commented:
**Scenario 2**

If signup with a valid email and username but an invalid password (I tried one with a single character), it advances to the next step momentarily then jumps back to the `user` step. But it's too late, I've been assigned a test group.

![Apr-24-2020 12-11-57](https://user-images.githubusercontent.com/1500769/80161613-1c019880-8625-11ea-9e9e-27754605e861.gif)

Sometime though it does validate the password before advancing. But I can't track down what the conditions are.

I don't know how important it is for you to fix this. It will muddy the collected stats. But I also think advancing to the next step too early is a long standing signup framework issue.

**Scenario 5**

My existing user was assigned to the `control` variation. The scenario description says I should expect the swap variation, but I'm not sure why this would be the case. Being assigned to `control` seems fine to me (although I haven't read pbKBsr-2c-p2 deeply)

**Scenario 7**

I think this copy needs changing when in the test variant:
> That's why we'll pay the registration fees for your new domain when you choose a paid plan during the next step.

The user no longer chooses the plan during the next step.

**Other**

I thought it'd be refresh the upsell page after updating my currency, but it took me back to the start of the flow. Generally I think it should be safe to refresh the page during the signup flow, but not a show stopper.

p-jackson commented on client/components/domains/free-domain-explainer/index.jsx:
```diff
@@ -20,8 +20,30 @@ class FreeDomainExplainer extends React.Component {
+		return (
+			<div className="free-domain-explainer card is-compact">
+				<header>
+					<h1 className="free-domain-explainer__title">All our custom domains need a paid plan.</h1>
+					<p className="free-domain-explainer__subtitle">
+						That's why we'll pay the registration fees for your new domain when you choose a paid
+						plan during the next step.
+					</p>
+					<p className="free-domain-explainer__subtitle">
+						You can always claim your free custom domain later if you aren't ready yet.
+					</p>
```
These strings look like they should be wrapped in `translate()`

niranjan-uma-shankar replied:
Not wrapping in translate is intentional - we don't know yet if this test will win, so we'd like to avoid translation costs until then.

p-jackson commented on client/lib/signup/step-actions/index.js:
```diff
@@ -740,3 +752,28 @@ export function isSiteTopicFulfilled( stepName, defaultDependencies, nextProps )
 	}
 }
+
+export function shouldShowUpsell( stepName, defaultDependencies, nextProps ) {
```
I always expect functions that start with `should` to be pure and probably to just return a boolean.

Maybe we could try renaming to `maybeShowUpsell()`? The name's a bit awkward, but all I could think of 😅

p-jackson replied:
Ah yes, it's used as one of those step fulfilled callbacks. The naming of these isn't very good across the board. Maybe `showOrHideUpsellStep()`?

niranjan-uma-shankar replied:
You're right with the naming. I'll push a change with the renamed function.

niranjan-uma-shankar replied:
Done in [75596ad](https://github.com/Automattic/wp-calypso/pull/40879/commits/75596ad43d43381391b65db799e083ee0367a740)

p-jackson commented on client/signup/config/flows.js:
```diff
@@ -163,7 +163,7 @@ const Flows = {
 	 */
 	excludeStep( step ) {
-		step && Flows.excludedSteps.push( step );
+		step && Flows.excludedSteps.indexOf( step ) === -1 && Flows.excludedSteps.push( step );
```
`excludedSteps` should probably be a `Set`. You don't have to do that refactor though, this is fine.

p-jackson commented on client/signup/steps/upsell/form.jsx:
```diff
@@ -0,0 +1,209 @@
+		return (
+			<header className="upsell__small-header">
+				<i className="upsell__title">
+					Your domain { this.props.domainName } is currently available
```
Some more strings here that aren't wrapped in `translate()`

niranjan-uma-shankar replied:
Please check [comment](https://github.com/Automattic/wp-calypso/pull/40879#discussion_r414280361).

p-jackson commented on client/state/signup/progress/actions.js:
```diff
@@ -128,3 +129,10 @@ export function invalidateStep( step, errors ) {
 	};
 }
+
+export function removeStep( step ) {
```
I may have missed it, but I can't see where this new action is used. I see it's passed into the the main `<Signup>` component, but don't see it get called. Is it still needed?

niranjan-uma-shankar replied:
I use removeStep in the `shouldShowUpsell()` function in [this line](https://github.com/Automattic/wp-calypso/pull/40879/files#diff-1236f5fcce1a6ae92a78305706a9d3f7R775). The reason removeStep is needed is that a user could do the following steps:
Select paid plan -> Hit back in domains step -> Select free plan -> Select custom domain.

Selecting a paid plan first will fulfill the `upsell-plan` and `plans-plan-only` steps and add those steps to the progress store as "completed". This is because we don't want to show the upsell for users who've selected a paid plan. Now, going back on the Domains step and selecting the free plan should remove the fulfilled steps from progress store since we might probably show the upsell depending on the path the user takes from here -> if they select a free domain, the steps upsell and second plan steps get fulfilled again; if they select a custom domain, then they see the upsell offer.

p-jackson replied:
That makes sense, thanks for explaining that.

The difference between `removeStep` (which removes the dependencies provided by the step) and "excluding a step" (which excludes it from the flow) is pretty subtle.

What do you think about calling it `unsubmitSignupStep()` since it's sort of goes hand in hand with `submitSignupStep()`? Or have I still misunderstood the purpose? 😅

niranjan-uma-shankar replied:
Agree. Going by the naming style followed in the [reducer file](https://github.com/Automattic/wp-calypso/blob/786ca0e353f64ccfd8d87fb538bbbbad1aca13a6/client/state/signup/progress/reducer.js#L111), I think unsubmitStep would work well. I'll do that 👍

p-jackson commented:
I'm out of time today, but code changes LGTM @niranjan-uma-shankar 👍 
I made those other comments above about the copy etc. happy for them to be resolved (or not) however you think is best.

niranjan-uma-shankar commented:
> Sometime though it does validate the password before advancing. But I can't track down what the conditions are.
> I don't know how important it is for you to fix this. It will muddy the collected stats. But I also think advancing to the next step too early is a long standing signup framework issue.

Oh boy, what a bug. I investigated this issue and it is certainly a signup framework issue. The framework will take you to the next step as soon as you submit the current step. However, if the current step has an API request in progress(like [creating a user account](https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/steps-pure.js#L146) in the case of the user step), it still takes us to the next step. If the API returns an error, then it will pull you back to the previous step([code ref](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/flow-controller.ts#L336)). This is why we see the flicker in your gif. 

I get why it might have been implemented this way. We don't want to get in the way of the user and make them wait for an API request to complete, so we are eager to take them to the next step. If the API for the previous step gives an error, it's probably okay to take them back and tell them about it. But, this messes up analytics and tracking, and other stuff. We already add the next step to the progress store; we fire tracks events for the next step like `calypso_signup_step_start`, `calypso_page_view`, `calypso_domain_search_pageview` etc. I also think we shouldn't be firing the `calypso_signup_actions_complete_step ` event in [this line](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/flow-controller.ts#L331) because the `if ( errors ) ...` check happens after that line, so if the account creation API for the user step errors out, the user step will still get marked "completed" and in the next line become an invalidated step. 
Most importantly, it messes up AB test assignments because we need to assign the AB test just before the step renders, and we can't know if the previous step will later fail(like it happens when we briefly see the domains step and then taken back to the user step).

I worked on a way to "fix"  this but it started getting complicated and I realized that it's probably not the best use of my time to fix a framework issue that's been around from the beginning, only for the purpose of an AB test that may or may not perform well. 

Without fixing the framework, we could make the user wait for API response on the user step and then assign the AB test. This will work and solve the issue, but I am not comfortable making a UX decision on adding more delay. Another thing we could do is to make the password validation happen before the account creation API gets called. This is something I have done in [aa30585](https://github.com/Automattic/wp-calypso/pull/40879/commits/aa3058551f610254cd5360e4fe4ece0c56323456), so at least password errors will get caught before AB test assignment, but there are other ways the account creation API could fail.

My preference is to go with this now, and normalize later via data analysis. My guess is that after the password fix I've made now, there won't be too many users who fail the user step.

niranjan-uma-shankar commented:
I have gone ahead and fixed the error in the `calypso_signup_actions_complete_step` tracks event firing - we will fire this tracks event only when there's no error in the API response. I think this will be helpful for data analysis of this AB test's result. Also, as I said before, in [aa30585](https://github.com/Automattic/wp-calypso/pull/40879/commits/aa3058551f610254cd5360e4fe4ece0c56323456) I have fixed the issue of password field not getting validated until form submit(the other fields don't behave this way, so no reason for the password field to be any different). 

@victorespigares The final summary is that we have a signup framework issue that could cause to assign the user to an AB test even when the user registration fails. Ideally we should wait assign to test only if user registration is successful but fixing it is IMO complicated and not worth pursuing. As a result, we might have some users who fail user registration, still get assigned to the test, and then drop off. In the post test analysis, we should not include such users in the stats. The number of users in `calypso_abtest_start` will include these false positives in both variations. A more reliable Hypothesis would be `calypso_signup_actions_complete_step` on the user step -> wpcom_product_purchase in both control and variation. This is because complete_step event is fired only when the account creation API has succeeded. 
My guess is that we won't see too many false positives in calypso_abtest_start so probably this is not a big deal, but it's better to be aware. We could run two Hypothesis, one with calypso_abtest_start and the other with calypso_signup_actions_complete_step on the user step and compare.

niranjan-uma-shankar commented:
@victorespigares  In the upsell offer copy, we have this line:

> Custom domains are only available on our paid plans, so you’ll need to upgrade to use <strong>helloworldgrey.com</strong> as your custom domain.

Please correct me if I'm wrong, but isn't this line no longer true with PWPO launched to all users earlier this week(check p7DVsv-8uh-p2) ? I believe custom domains are availabel to free users as a non-primary domain.

---
If the user selects a free plan and then selects to transfer or map a domain, I think we might need some copy changes in a few places. Picking a few lines:
- The small header text`Your domain helloworldgrey.com is currently available`: If it's a map or transfer domain, then the domain is already owned by the user.
- The main header text - `Upgrade to a paid plan to claim your custom domain for free`: Should this be "map your custom domain for free" or "transfer your custom domain for free"?
- `Custom domains are only available on our paid plans`: Should this be mapped domains or transfer domains?
- `That’s why we’ll cover your domain registration for the first year`: Should this be "cover your domain mapping" or "domain transfer" cost?

victorespigares commented:
@niranjan-uma-shankar good points with the copy. I pushed some modifications in https://github.com/Automattic/wp-calypso/pull/40879/commits/b00a38b0316eff8d5e7610463335cba22c56ab0f to include those cases.

p-jackson commented:
> My preference is to go with this PR now, and normalize later via data analysis. My guess is that after the password fix I've made now, there won't be too many users who fail the user step.

I agree, it'd would be a lot to invest in a single test.

p-jackson commented on client/signup/utils.js:
```diff
@@ -11,15 +11,39 @@ import { translate } from 'i18n-calypso';
+function isEligibleForSwapStepsTest() {
+	const cookies = cookie.parse( document.cookie );
+	const countryCodeFromCookie = cookies.country_code;
+	const isUserFromUS = 'US' === countryCodeFromCookie;
```
It looks like the abtest library might already have a way to limit the test to a particular country using the `countryCodeTargets` property described in the readme: https://github.com/Automattic/wp-calypso/blob/master/client/lib/abtest/README.md
